src/services/encryptionService.js

```diff
@@ -1,38 +1,130 @@
+    await localforage.removeItem("account");
+  }
+
+  async #getAccount(lockPassword) {
+    if (this.#account) {
+      return this.#account;
     }
 
-    this.account = new Account();
+    const key = (await createHash(lockPassword)).slice(0, 32);
+    const encAuthKey = await localforage.getItem("account");
 
-    return this.account;
+    let isNewAccount = false;
+    if (encAuthKey) {
+      try {
+        this.#account = Account.from_pickle(encAuthKey, key);
+      } catch (error) {
+        showCustomAlert("Incorrect key. Please try again.");
+        throw new Error("[encryption] Incorrect key.");
+      }
+    } else {
+      this.#account = new Account();
+      isNewAccount = true;
+      const encryptedKey = this.#account.pickle(key);
+      localforage.setItem("account", encryptedKey);
+    }
+    return { account: this.#account, isNewAccount };
   }
 
-  async registerDevice() {
-    const user = this.#getAccount();
-    user.generate_one_time_keys(50);
+  async registerDevice(lockPassword) {
+    const { account, isNewAccount } = await this.#getAccount(lockPassword);
 
-    setEncryptedUser(user);
+    if (!account) return;
+    if (!isNewAccount) return { isSuccessAuth: true };
+
+    account.generate_one_time_keys(50);
+    console.log("Encryption: User account generated", account);
 
     const data = {
-      identity_key: user.curve25519_key,
-      signed_key: user.ed25519_key,
-      one_time_pre_keys: Object.fromEntries(user.one_time_keys.entries()),
+      identity_key: account.curve25519_key,
+      signed_key: account.ed25519_key,
+      one_time_pre_keys: Object.fromEntries(account.one_time_keys.entries()),
     };
-    await api.encryptedDeviceCreate(data);
+
+    try {
+      await api.encryptedDeviceCreate(data);
+      return { isSuccessAuth: true };
+    } catch (error) {
+      console.error("[encryption] Failed to register device", error);
+      return { isSuccessAuth: false };
+    }
+  }
+
+  async #getUserKeys(userId) {
+    try {
+      const usersKeys = await api.getEncryptedKeys({ user_ids: [userId] });
+      const keys = usersKeys[userId]?.[0];
+
+      if (keys) {
+        store.dispatch(upsertUser({ _id: userId, keys }));
+      } else {
+        console.error("[encryption] No keys found for the user", userId);
+      }
+
+      return keys;
+    } catch (error) {
+      throw new Error("[encryption] Failed to get participant's keys");
+    }
+  }
+
+  async createEncryptionSession(userId) {
+    if (!this.#vodozemacInitialized || !this.#account) {
+      throw new Error("[encryption] Service or account not initialized");
+    }
+
+    const userKeys = await this.#getUserKeys(userId);
+    if (!userKeys) {
+      throw new Error("[encryption] Could not retrieve opponent's keys");
+    }
+
+    try {
+      const session = this.#account.create_outbound_session(
+        userKeys.identity_key,
+        userKeys.one_time_pre_keys
+      );
+      this.#encryptionSession = session;
```

Comment:
there will be a way to have multiple active e2ee chats. Let's have `this.#encryptionSession  ` as a map and store sessions like `this.#encryptionSession[userId] = session `